nntrainer/layers/multi_head_attention_layer.cpp

```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
+  NNTR_THROW_IF(provide_attention_mask && mask_dim.channel() != num_heads,
+                std::invalid_argument)
+    << "Dimension of input mask channel: " << mask_dim.channel()
+    << " is not matched with num_heads property: " << num_heads << " for layer "
+    << context.getName();
+
+  NNTR_THROW_IF(key_height != value_height, std::invalid_argument)
+    << "Dimension of input key height: " << key_height
+    << " is not matched with Dimension of input value height: " << value_height
+    << " for layer " << context.getName();
+  NNTR_THROW_IF(provide_attention_mask && mask_dim.height() != query_height,
+                std::invalid_argument)
+    << "Dimension of input mask height: " << mask_dim.height()
+    << " is not matched with Dimension of input query height: " << query_height
+    << " for layer " << context.getName();
+
+  NNTR_THROW_IF(provide_attention_mask && mask_dim.width() != key_height,
+                std::invalid_argument)
+    << "Dimension of input mask width: " << mask_dim.width()
+    << " is not matched with Dimension of input key height: " << key_height
+    << " for layer " << context.getName();
+
+  if (std::get<props::ProjectedKeyDim>(multi_head_attention_props).empty()) {
+    NNTR_THROW_IF(query_width % num_heads, std::invalid_argument)
+      << "query_width: " << query_width
+      << " is not divisible by num_heads: " << num_heads << " for layer "
+      << context.getName();
+
+    ml_logw("[multi head attention] ProjectedKeyDim property is not given. "
+            "Default value(query_width / num_heads) is set");
+
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props)
+      .set(query_width / num_heads);
+  }
+  const unsigned int projected_key_dim_prop =
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props).get();
+
+  if (std::get<props::ProjectedValueDim>(multi_head_attention_props).empty()) {
+    std::get<props::ProjectedValueDim>(multi_head_attention_props)
+      .set(projected_key_dim_prop);
+  }
+
+  if (std::get<props::OutputShape>(multi_head_attention_props).empty()) {
+    std::get<props::OutputShape>(multi_head_attention_props).set(query_width);
+  }
+
+  sm.setActiFunc(ActivationType::ACT_SOFTMAX);
+}
 
 void MultiHeadAttentionLayer::finalize(InitLayerContext &context) {
   finalizeCommon(context);
+
+  const bool disable_bias =
+    std::get<props::DisableBias>(*layer_impl_props).get();
+  auto &weight_initializer =
+    std::get<props::WeightInitializer>(*layer_impl_props).get();
+  auto &bias_initializer = std::get<props::BiasInitializer>(*layer_impl_props);
+  auto &weight_regularizer =
+    std::get<props::WeightRegularizer>(*layer_impl_props);
+  auto &weight_regularizer_constant =
+    std::get<props::WeightRegularizerConstant>(*layer_impl_props);
+  const float &weight_decay =
+    std::get<props::WeightDecay>(*layer_impl_props).get();
+  const float &bias_decay = std::get<props::BiasDecay>(*layer_impl_props).get();
```

Comment:
How about move these layer_impl property to `finalizeCommon` if it should exist.

Reply:
Done